wiki/python/DEVELOPER_GUIDE.md

```diff
@@ -47,31 +47,110 @@ The examples in the documentation are checked through doctest. To skip the check
-1. Every test has at least one deterministic test case that always runs
-2. Critical edge cases are documented and tested consistently
-3. Test failures can be reproduced reliably
-
-Note: While the explicit examples will always run in CI, the hypothesis-generated test cases (from `@given` strategies) only run during nightly testing by default. This ensures fast CI runs while still maintaining thorough testing coverage.
-
-Example of a valid hypothesis test:
-```python
-@example(dtype=np.float32, sparse_input=False)  # baseline case, runs as part of PR CI
-@example(dtype=np.float64, sparse_input=True)   # edge case, runs as part of PR CI
-@given(
-    dtype=st.sampled_from((np.float32, np.float64)),
-    sparse_input=st.booleans()
-)  # strategy-based cases, only runs during nightly tests
-def test_my_estimator(dtype, sparse_input):
-    # Test implementation
-    pass
+We use [pytest](https://docs.pytest.org/en/latest/) for writing and running tests. To see existing examples, refer to any of the `test_*.py` files in the folder `cuml/tests`.
+
+### Test Organization
+- Keep all tests for a single estimator in one file, with exceptions for:
+  - Performance testing/benchmarking
+  - Generic estimator checks (e.g., `test_base.py`)
+- Use small, focused datasets for correctness testing
+- Only parametrize scale when it triggers alternate code paths
+
+### Test Input Generation
+We support three main approaches for test input generation:
+
+1. **Fixtures** (`@pytest.fixture`):
+   - For shared setup/teardown code and resources
+   - Examples: random seeds, clients, loading test datasets
+   ```python
+   @pytest.fixture(scope="module")
+   def random_state():
+       return 42
+   ```
+
+2. **Parametrization** (`@pytest.mark.parametrize`):
+   - For testing specific input combinations
+   - Good for hyperparameters and configurations
+   ```python
+   @pytest.mark.parametrize("solver", ["svd", "eig"])
+   def test_estimator(solver):
+       pass
+   ```
+
+3. **Hypothesis** (`@given`):
+   - For property-based testing with random inputs
+   - Must include at least one `@example` for deterministic testing
+   - Preferred for dataset generation
+   ```python
+   @example(dataset=small_regression_dataset(np.float32))
+   @given(dataset=standard_regression_datasets())
+   def test_estimator(dataset):
+       pass
+   ```
+
+### Test Parameter Levels
+
+You can mark test parameters for different scales with (`unit_param`, `quality_param`, and `stress_param`).
+
+_Note: For dataset scaling, prefer using hypothesis, e.g. with `standard_regression_datasets()`._
+
+We provide three test parameter levels:
+
+1. **Unit Tests** (`unit_param`): Small values for quick, basic functionality testing
+   ```python
+   unit_param(2)  # For number of components
+   ```
+
+2. **Quality Tests** (`quality_param`): Medium values for thorough testing
```

Comment:
Referring to my last comment, `medium` is not well defined here, as well as the difference between basic vs thorough.

Reply:
See my above comment, maybe we can take a step back and think about the motivation for this, and that might help us to determine whether we need this at all, and if yes, how exactly we want to define this?